code/modules/research/techweb/_techweb.dm

```diff
@@ -0,0 +1,272 @@
+		research_node(TN, TRUE)
+	hidden_nodes = list()
+
+/datum/techweb/science	//Global science techweb for RND consoles.
+	id = "SCIENCE"
+
+/datum/techweb/Destroy()
+	researched_nodes = null
+	researched_designs = null
+	available_nodes = null
+	visible_nodes = null
+	return ..()
+
+/datum/techweb/proc/recalculate_nodes(recalculate_designs = FALSE)
+	var/list/datum/techweb_node/processing = list()
+	for(var/i in researched_nodes)
+		processing[i] = researched_nodes[i]
+	for(var/i in visible_nodes)
+		processing[i] = visible_nodes[i]
+	for(var/i in available_nodes)
+		processing[i] = available_nodes[i]
+	for(var/i in processing)
+		update_node_status(processing[i])
+	if(recalculate_designs)					//Wipes custom added designs like from design disks or anything like that!
+		researched_designs = list()
+	for(var/i in processing)
+		var/datum/techweb_node/TN = processing[i]
+		if(researched_nodes[TN.id])
+			for(var/I in TN.designs)
+				researched_designs[I] = TN.designs[I]
+
+/datum/techweb/proc/copy_research_to(datum/techweb/reciever, unlock_hidden = TRUE)				//Adds any missing research to theirs.
+	for(var/i in researched_nodes)
+		reciever.researched_nodes[i] = researched_nodes[i]
+	for(var/i in researched_designs)
+		reciever.researched_designs[i] = researched_designs[i]
+	if(unlock_hidden)
+		for(var/i in reciever.hidden_nodes)
+			if(!hidden_nodes[i])
+				reciever.hidden_nodes -= i		//We can see it so let them see it too.
+	reciever.recalculate_nodes()
+
+/datum/techweb/proc/copy()
+	var/datum/techweb/returned = new()
+	returned.researched_nodes = researched_nodes.Copy()
+	returned.visible_nodes = visible_nodes.Copy()
+	returned.available_nodes = available_nodes.Copy()
+	returned.researched_designs = researched_designs.Copy()
+	returned.hidden_nodes = hidden_nodes.Copy()
+	return returned
+
+/datum/techweb/proc/get_visible_nodes()			//The way this is set up is shit but whatever.
+	return visible_nodes - hidden_nodes
+
+/datum/techweb/proc/get_available_nodes()
+	return available_nodes - hidden_nodes
+
+/datum/techweb/proc/get_researched_nodes()
+	return researched_nodes - hidden_nodes
+
+/datum/techweb/proc/add_design_by_id(id)
+	return add_design(get_techweb_design_by_id(id))
+
+/datum/techweb/proc/add_design(datum/design/design)
+	if(!istype(design))
+		return FALSE
+	researched_designs[design.id] = design
+	return TRUE
+
+/datum/techweb/proc/remove_design_by_id(id)
+	return remove_design(get_techweb_design_by_id(id))
+
+/datum/techweb/proc/remove_design(datum/design/design)
+	if(!istype(design))
+		return FALSE
+	researched_designs[design.id] = design
+	return TRUE
+
+/datum/techweb/proc/research_node_id(id, force, auto_update_points)
+	return research_node(get_techweb_node_by_id(id), force, auto_update_points)
+
+/datum/techweb/proc/research_node(datum/techweb_node/node, force = FALSE, auto_adjust_cost = TRUE)
+	if(!istype(node))
+		return FALSE
+	update_node_status(node)
+	if(!force)

```

Comment:
This fail if someone adds node with a cost to starting ones, which seems plausible ?

Reply:
fixed